### PR TITLE
N7S ETB#97 create compiler wrapper

### PR DESCRIPTION
**Pull request by Lurkerpas**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+## along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+SRC_DIR=src
+
+.PHONY: all test adac rt clean
+
+all: adac rt test
+
+adac:
+	( ${MAKE} -C ${SRC_DIR}/arm-elf-adac) || exit $$?;
```
Comment (rphh) [inline Makefile:23]: Is this `|| exit $$?;` required? If children make fails then the parent fails too

File: src/arm-elf-adac/Makefile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+##
+## This program is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+TOP_DIR=$(shell realpath ../../)
```
Comment (rphh) [inline src/arm-elf-adac/Makefile:16]: Is this used?

File: src/arm-elf-adac/arm-elf-adabind [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import os
+
+
+class BexchData:
```
Comment (rphh) [inline src/arm-elf-adac/arm-elf-adabind:24]: I have a feeling that I wrote this code. This requires doc...

File: src/arm-elf-adac/arm-elf-adac [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import os
+import tempfile
+
+ADA_DEF_EXTENSION = ".adb"
```
Comment (rphh) [inline src/arm-elf-adac/arm-elf-adac:22]: documentation?

File: src/gnat-runtime/Common.mk [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+%.o: %.c
+	$(ARM_CC) $(ARM_CFLAGS) -c $< -o $@
+
+${GNAT_CRT_ARM}: ${OBJ} ${C_OBJ}
+	${AR} cr $@ ${OBJ} ${C_OBJ}
+
+clean:
+	rm -f *.o
+	rm -f *.ali
+	rm -f ${GNAT_CRT_ARM}
```
Comment (rphh) [inline src/gnat-runtime/Common.mk]: No newline at the end of file

File: src/gnat-runtime/Makefile [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+clean:
+ifneq ("$(wildcard $(ARM_OBJ_DIR)/Makefile)","")
+	${MAKE} -C ${ARM_OBJ_DIR} clean
+endif
+ifneq ("$(wildcard $(ARM_OBJ_SF_DIR)/Makefile)","")
+	${MAKE} -C ${ARM_OBJ_SF_DIR} clean
+endif
+	rm -r -f ${BUILD_DIR}
+	rm -r -f ${LIB_DIR}
+	rm -r -f ${INC_DIR}
```
Comment (rphh) [inline src/gnat-runtime/Makefile]: No newline at the end of file

File: test/Makefile [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	cinterop \
+	floatmath \
+	gpr \
+	intmath \
+	opengeode/basicsdl \
+	opengeode/managemode
+	
+
+test:
+	for t in ${TESTS}; do \
```
Comment (rphh) [inline test/Makefile]: This can be refactored: https://stackoverflow.com/questions/17834582/run-make-in-each-subdirectory

File: test/adainterop/main.c [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    const int32_t a2 = 64561;
+    const int32_t a3 = 15632;
+    const int32_t a4 = 52345;
+    const int32_t result = addMany(a1, a2, a3, a4);
+    if (result == (a1 + a2 + a3 + a4))
+        return true;
+    return false;
+}
+
+void runTests(void)
```
Comment (rphh) [inline test/adainterop/main.c:84]: From the perspective of testing: corner cases and equivalence classes:
0 + 0
0 + -10
0 + 10
-10 + 0
10 + 0
-10 + 20
20 + -10
maxint + maxint
maxint + 0
0 + maxint
minint + minint
minint + 0
0 + minint
minint + minint
maxint + minint
minint + maxint

File: test/adainterop/main.map [hunk 9]
```diff
@@ -0,0 +1,1465 @@
+Archive member included to satisfy reference by file (symbol)
```
Comment (rphh) [inline test/adainterop/main.map]: I guess that this file shouldn't be in the repo

File: test/ads/bodylessmodule.ads [hunk 10]
```diff
@@ -0,0 +1,9 @@
+with utils;
+use utils;
+
+package BodylessModule is
+
+function functionInHeader(a : int32_t) return int32_t is (a * a + 1);
+pragma Export(C, functionInHeader, "functionInHeader");
+
+end BodylessModule;
```
Comment (rphh) [inline test/ads/bodylessmodule.ads]: Missing newline

File: test/ads/main.c [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    reporting__reportsuccess();
+}
+
+int main(void)
+{
+    runTests();
+    for (;;);
+    return 0;
+}
```
Comment (rphh) [inline test/ads/main.c]: no newline

File: test/asn1scc/rtl/acn_asn1_rtl.adb [hunk 12]
```diff
@@ -0,0 +1,1381 @@
+with Ada.Unchecked_Conversion;
```
Comment (rphh) [inline test/asn1scc/rtl/acn_asn1_rtl.adb:1]: Maybe this repo shall include asn1scc as submodule?

File: test/gpr/GprBuildBasedProject.gpr [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                & SAMV71_Switches;
+      end case;
+
+   end Linker;
+
+   package Builder is
+      for Executable ("main.c") use "gpr.elf";
+   end Builder;
+
+end GprBuildBasedProject;
```
Comment (rphh) [inline test/gpr/GprBuildBasedProject.gpr]: missing newline

**Review by rphh — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (Lurkerpas) [reply Makefile:23]: Yes, that is the desired outcome

File: src/arm-elf-adac/Makefile (hunk 2 above)
Comment (Lurkerpas) [reply src/arm-elf-adac/Makefile:16]: Yes, in BIN_DIR, which is one line below... and which is used in the formulas.
BIN_DIR is not assembled directly, as argument to realpath must exist...

File: src/arm-elf-adac/arm-elf-adabind (hunk 3 above)
Comment (Lurkerpas) [reply src/arm-elf-adac/arm-elf-adabind:24]: Yes, this is your code.
Will try to provide docs.

Comment (Lurkerpas) [reply src/arm-elf-adac/arm-elf-adabind:24]: Done

File: src/gnat-runtime/Common.mk (hunk 5 above)
Comment (Lurkerpas) [reply src/gnat-runtime/Common.mk]: Fixed

File: test/adainterop/main.map (hunk 9 above)
Comment (Lurkerpas) [reply test/adainterop/main.map]: Fixed

File: test/ads/bodylessmodule.ads (hunk 10 above)
Comment (Lurkerpas) [reply test/ads/bodylessmodule.ads]: Fixed

File: test/ads/main.c (hunk 11 above)
Comment (Lurkerpas) [reply test/ads/main.c]: Fixed

File: test/asn1scc/rtl/acn_asn1_rtl.adb (hunk 12 above)
Comment (Lurkerpas) [reply test/asn1scc/rtl/acn_asn1_rtl.adb:1]: I don't think that this is a good idea. Little benefit, while attaching a quite large repository, with external instabilities.

File: test/gpr/GprBuildBasedProject.gpr (hunk 13 above)
Comment (Lurkerpas) [reply test/gpr/GprBuildBasedProject.gpr]: Fixed

File: src/arm-elf-adac/arm-elf-adac (hunk 4 above)
Comment (Lurkerpas) [reply src/arm-elf-adac/arm-elf-adac:22]: Done

File: src/gnat-runtime/Common.mk (hunk 5 above)
Comment (rphh) [reply src/gnat-runtime/Common.mk]: github still shows the warning

Comment (Lurkerpas) [reply src/gnat-runtime/Common.mk]: There was a tab....

File: src/gnat-runtime/Makefile (hunk 6 above)
Comment (Lurkerpas) [reply src/gnat-runtime/Makefile]: fixed

File: test/adainterop/main.c (hunk 8 above)
Comment (Lurkerpas) [reply test/adainterop/main.c:84]: The purpose of this is not to test the actual arithmetic, but handling of integers of different sizes. This actually detected issues with 16-bit ints on MSP430.

File: test/Makefile (hunk 7 above)
Comment (Lurkerpas) [reply test/Makefile]: But I want to explicitly list the tests... First, not all directories contain tests... secondly, asn1scc tests are disabled

Comment (rphh) [reply test/Makefile]: Ok, keep it like that

**Review by rphh — APPROVED** [review]
